sourcetracker/_sourcetracker.py

```diff
@@ -583,16 +634,109 @@ def gibbs_sampler(sink, cp, restarts, draws_per_restart, burnin, delay):
                 final_env_assignments[drawcount] = seq_env_assignments
                 final_taxon_assignments[drawcount] = taxon_sequence
 
-                # We've made a draw, update this index so that the next
-                # iteration will be placed in the correct index of results.
+                if autoc is not None:
+                    converged = autoc(final_envcounts, autoc_drawcount,
+                                      drawcount)
+                    if converged:
+                        convergence_times[restart, 0] = drawcount
+                        convergence_times[restart, 1] = autoc_drawcount
+                        break
+                    elif not converged and rep < total_passes:
+                        autoc_drawcount += 1
+                    elif not converged and rep == total_passes:
+                        convergence_times[restart, 0] = drawcount
+                        convergence_times[restart, 1] = -1
+
+                # We've completed making a draw (and checking for convergence
+                # if asked to do so). We incriment the index so that the next
+                # iteration will be placed in the correct row of results.
                 drawcount += 1
 
-    return (final_envcounts, final_env_assignments, final_taxon_assignments)
+    if autoc is not None:
+        conv_chain_fraction = (convergence_times[:, 1] != -1).sum() / restarts
+
+        # For chains that converged we'll return the first draw satisfying the
+        # convergence criterion. For chains that don't converge we'll return
+        # the final draw (`draws_per_restart`).
+        idxs = convergence_times[:, 0]
+        final_envcounts = final_envcounts[idxs, :]
+        final_env_assignments = final_env_assignments[idxs, :]
+        final_taxon_assignments = final_taxon_assignments[idxs, :]
+
+        times = np.where(convergence_times[:, 1] == -1, np.nan,
+                         convergence_times[:, 1])
+        return (final_envcounts, final_env_assignments,
+                final_taxon_assignments, times, conv_chain_fraction)
+
+    else:
```

Comment:
You really shouldn't have different collections of values in the return here, it makes code that uses this much harder to work with as the calling function has to anticipate what it's going to get back. It would be much better to return an empty `np.array` in place of `times`, and a float that makes sense (0.0?) in place of `conv_chain_fraction`. If you don't want to do that, you should split this entire function into two different functions corresponding to the different behavior. For example, `gibbs_sampler` and `gibbs_sampler_autoc`.